storage-node/src/storage_manager.rs

```diff
@@ -13,24 +16,142 @@ use crate::{
 /// We should allow concurrent requests fed into the storage manager,
 /// which should be responsible for handling multiple requests at the
 /// same time.
+
 pub struct StorageManager<C: ParpulseCache> {
-    // TODO: Consider making the cache lock-free. See the comments for
-    // `ParpulseCache`.
-    cache: Arc<RwLock<C>>,
+    // TODO: Consider making the cache lock-free.
+    cache: RwLock<C>,
+    disk_manager: Mutex<DiskManager>,
+    // TODO: cache_base_path should be from config
+    // Cache_key = S3_PATH; Cache_value = (CACHE_BASE_PATH + S3_PATH, size)
+    cache_base_path: String,
 }
 
 impl<C: ParpulseCache> StorageManager<C> {
-    pub fn new(cache: C) -> Self {
+    pub fn new(cache: C, disk_manager: DiskManager, cache_base_path: String) -> Self {
         Self {
-            cache: Arc::new(RwLock::new(cache)),
+            cache: RwLock::new(cache),
+            disk_manager: Mutex::new(disk_manager),
+            cache_base_path,
         }
     }
 
-    pub async fn get_data(&self, _request: RequestParams) -> StorageResult<()> {
-        // TODO:
+    pub async fn get_data(&self, request: RequestParams) -> StorageResult<usize> {
         // 1. Try to get data from the cache first.
         // 2. If cache miss, then go to storage reader to fetch the data from
         // the underlying storage.
-        todo!()
+        // 3. If needed, update the cache with the data fetched from the storage reader.
+
+        // TODO: Support more request types.
+        let key = request.as_file().unwrap();
+        let mut cache = self.cache.write().await;
+
+        match cache.get(key) {
+            Some(value) => {
+                // Directly get data from the cache.
+                let cache_value_size = value.size();
+                let cache_file_path = value.as_value();
+
+                let mut output_iterator = self
+                    .disk_manager
+                    .lock()
+                    .await
+                    .disk_read_sync_iterator(cache_file_path, 1024)?;
+                let mut data_size: usize = 0;
+                loop {
```

Comment:
Maybe we can put this loop out of match cache.get(key) for now? Since currently I use this loop to respond!

Reply:
If we read the data from the underlying storage (e.g. S3), ideally the data should be written into the response buffer from the storage buffer. The code seems to first put the data into the disk cache and then read from the disk cache, which incurs some unnecessary IO. I was thinking we may need to change this and there're a bunch of other features to add anyway, so I just moved the loop into one of the branches. 🥹

Reply:
Make sense! Great thinking!